File: src/main/java/pro/sky/java/weatherapp/service/OwmWeatherService.java
package pro.sky.java.weatherapp.service;

import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import pro.sky.java.weatherapp.component.OwmRestClient;
import pro.sky.java.weatherapp.domain.HistoryRecord;
import pro.sky.java.weatherapp.domain.Weather;
import reactor.core.publisher.Mono;

@Slf4j
@Service
public class OwmWeatherService implements WeatherService {

    private final OwmRestClient restClient;
    private final HistoryService historyService;

    public OwmWeatherService(
            OwmRestClient restClient,
            HistoryService historyService
    ) {
        this.restClient = restClient;
        this.historyService = historyService;
    }

    @Override
    @SneakyThrows
    public Mono<Weather> getWeatherByCity(String cityName) {
        log.info("Trying to retrieve weather forecast: {}", cityName);
        return restClient.getForecast(cityName)
                .doOnSuccess(f ->
                        historyService.save(new HistoryRecord(cityName, f))
                );
    }
}

File: src/main/kotlin/pro/sky/java/weatherapp/service/OwmWeatherService.kt
package pro.sky.java.weatherapp.service

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import mu.KotlinLogging
import org.springframework.stereotype.Service
import pro.sky.java.weatherapp.component.OwmRestClient
import pro.sky.java.weatherapp.domain.HistoryRecord
import pro.sky.java.weatherapp.domain.Weather

@Service
class OwmWeatherService(
    private val restClient: OwmRestClient,
    private val historyService: HistoryService
) : WeatherService {

    private val log = KotlinLogging.logger { }

    override suspend fun getWeatherByCity(cityName: String): Weather {
        log.info("Trying to get a forecast for {}", cityName)
        return restClient.getForecast(cityName).also {
            saveHistoryRecord(HistoryRecord(cityName, it))
        }
    }

    private suspend fun saveHistoryRecord(record: HistoryRecord) {
        withContext(Dispatchers.IO) {
            launch {
                historyService.save(record)
            }
        }
    }
}